src/debugger.rs

```diff
@@ -0,0 +1,218 @@
+use crate::Context;
+use crate::ContextPeopleExt;
+use crate::IxaError;
+use clap::{ArgMatches, Command};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::io::Write;
+
+trait DebuggerCommand {
+    /// Handle the command and any inputs; returning true will exit the debugger
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        matches: &ArgMatches,
+    ) -> Result<bool, String>;
+    fn about(&self) -> &'static str;
+}
+
+struct DebuggerRepl {
+    commands: HashMap<&'static str, Box<dyn DebuggerCommand>>,
+    output: RefCell<Box<dyn Write>>,
+}
+
+fn flush() -> Result<(), String> {
+    std::io::stdout().flush().map_err(|e| e.to_string())
+}
+
+impl DebuggerRepl {
+    fn new() -> Self {
+        DebuggerRepl {
+            commands: HashMap::new(),
+            output: RefCell::new(Box::new(std::io::stdout())),
+        }
+    }
+
+    fn register_command(&mut self, name: &'static str, handler: Box<dyn DebuggerCommand>) {
+        self.commands.insert(name, handler);
+    }
+
+    fn get_command(&self, name: &str) -> Option<&dyn DebuggerCommand> {
+        self.commands.get(name).map(|command| &**command)
+    }
+
+    fn writeln(&self, formatted_string: &str) -> Result<(), String> {
```

Comment:
I don't think this function should fail even in principle.. There's nothing reasonable for the command to do if it does.